OpenRA.Game/GlobalChat.cs

```diff
@@ -0,0 +1,353 @@
+
+		public GlobalChat()
+		{
+			client = new IrcClient();
+			client.Encoding = System.Text.Encoding.UTF8;
+			client.SendDelay = 100;
+			client.ActiveChannelSyncing = true;
+
+			client.OnConnecting += OnConnecting;
+			client.OnConnected += OnConnected;
+			client.OnDisconnecting += OnDisconnecting;
+			client.OnDisconnected += OnDisconnected;
+			client.OnError += OnError;
+			client.OnKick += OnKick;
+
+			client.OnRawMessage += (_, e) => Game.RunAfterTick(() => Log.Write("irc", e.Data.RawMessage));
+			client.OnJoin += OnJoin;
+			client.OnChannelActiveSynced += OnChannelActiveSynced;
+			client.OnTopic += (_, e) => topic = e.Topic;
+			client.OnTopicChange += (_, e) => topic = e.NewTopic;
+			client.OnNickChange += OnNickChange;
+
+			client.OnChannelMessage += (_, e) => AddMessage(e.Data.Nick, e.Data.Message);
+			client.OnOp += (_, e) => SetUserOp(e.Whom, true);
+			client.OnDeop += (_, e) => SetUserOp(e.Whom, false);
+			client.OnVoice += (_, e) => SetUserVoiced(e.Whom, true);
+			client.OnDevoice += (_, e) => SetUserVoiced(e.Whom, false);
+			client.OnPart += OnPart;
+			client.OnQuit += OnQuit;
+		}
+
+		void SetUserOp(string whom, bool isOp)
+		{
+			Game.RunAfterTick(() =>
+			{
+				ChatUser user;
+				if (Users.TryGetValue(whom, out user))
+					user.IsOp = isOp;
+			});
+		}
+
+		void SetUserVoiced(string whom, bool isVoiced)
+		{
+			Game.RunAfterTick(() =>
+			{
+					ChatUser user;
+				if (Users.TryGetValue(whom, out user))
+					user.IsVoiced = isVoiced;
+			});
+		}
+
+		public void Connect()
+		{
+			if (client.IsConnected)
+				return;
+
+			try
+			{
+				client.Connect(Game.Settings.Chat.Hostname, Game.Settings.Chat.Port);
+			}
+			catch (Exception e)
+			{
+				connectionStatus = ChatConnectionStatus.Error;
+				AddNotification(e.Message);
+				Game.RunAfterTick(() => Log.Write("irc", e.ToString()));
+			}
+
+			new Thread(client.Listen) { Name = "IrcListenThread" }.Start();
+		}
+
+		void AddNotification(string text)
+		{
+			var message = new ChatMessage(DateTime.Now, ChatMessageType.Notification, null, text);
+			Game.RunAfterTick(() =>
+			{
+				History.Add(message);
+				Log.Write("irc", text);
+			});
+		}
+
+		void AddMessage(string nick, string text)
+		{
+			var message = new ChatMessage(DateTime.Now, ChatMessageType.Message, nick, text);
+			Game.RunAfterTick(() =>
+			{
+				History.Add(message);
+				Log.Write("irc", text);
+			});
+		}
+
+		void OnConnecting(object sender, EventArgs e)
+		{
+			AddNotification("Connecting to {0}:{1}...".F(Game.Settings.Chat.Hostname, Game.Settings.Chat.Port));
+			connectionStatus = ChatConnectionStatus.Connecting;
+		}
+
+		void OnConnected(object sender, EventArgs e)
+		{
+			AddNotification("Connected.");
+			connectionStatus = ChatConnectionStatus.Connected;
+
+			// Guard against settings.yaml modification
+			Game.Settings.Chat.Nickname = SanitizedName(Game.Settings.Chat.Nickname);
```

Comment:
This write isn't thread-safe (since we usually only muck with globals on the UI thread).